### PR TITLE
Item hover outline FX

### PR DESCRIPTION
**Pull request by maxymax13**
## About The Pull Request

Adds feedback when you hover over stuff, from BeeStation.
Now with a pref and theme support!
dreamseeker_AJPobAoUme
dreamseeker_UfQv9Oyd2n
dreamseeker_61fgXjVB8D
dreamseeker_9nmfnPnBrQ
dreamseeker_9njKh3z6h5
dreamseeker_OUKctoxHhv
image

TODO:

- [x] Make the pref to not anger boomers

## Why It's Good For The Game
Feedback for when you're actually hovering over an item is nice, also helps you see dark items in inventory slots better.

## Changelog
:cl:
add: Items in your inventory now have an outline when you hover over them, originally from BeeStation. This can be disabled in Game Preferences.
/:cl:

### CODE REVIEW

**Comment by Qustinnus** [comment]
this only working in inv is kind of dissapointing and inconsistent.

**Comment by maxymax13** [comment]
> 
> 
> this only working in inv is kind of dissapointing and inconsistent.

how so? also, is there even a way of sending an outline effect to only one client

**Comment by Qustinnus** [comment]
the inventory is one of the only places where an outline is not very useful/relevant because the information/shapes are way easier to discern due to it being structured in your inventory rather than being tossed on a pile on the floor (Where the shape might be obscured by other items).

Doing this client-only is probably possible but its not as easy as doing it like this.

**Comment by maxymax13** [comment]
you can just shift rmb if you want something out of a pile, and having hover effects for everything like on ss14 would be cool but that's not what the scope for this PR is at all, this is for simple ui feedback on for example if your click will select the slot or select the item, stuff like that

**Comment by TheBonded** [comment]
Doesn't this already happen? Whenever I hover over things for a second, they give me a bit of information and tell me the strength (ie. Exceptionally robust, Very robust, Robust, Medium, Low, Very low). it also changes with the UI theme I chose. What exactly is this PR changing?

**Comment by Qustinnus** [comment]
it adds an outline filter

**Comment by TheBonded** [comment]
> it adds an outline filter

Ohhh, I see. Looks nice!

File: code/__DEFINES/colors.dm [hunk 1]
```diff
@@ -104,6 +104,15 @@
...
 #define COLOR_SAMPLE_GRAY "#5e5856"
 
+///Main colors for UI themes
+#define COLOR_THEME_MIDNIGHT	"#6086A0"
+#define COLOR_THEME_PLASMAFIRE	"#FFB200"
+#define COLOR_THEME_RETRO		"#24CA00"
+#define COLOR_THEME_SLIMECORE	"#4FB259"
+#define COLOR_THEME_OPERATIVE	"#B8221F"
+#define COLOR_THEME_GLASS		"#75A4C4"
+#define COLOR_THEME_CLOCKWORK	"#CFBA47"
```
Comment (Mothblocks) [inline code/__DEFINES/colors.dm:114]: No indentation alignment.

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#tabs-not-spaces

File: code/game/objects/items.dm [hunk 2]
```diff
@@ -191,6 +191,9 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
 	var/list/juice_results
 
 	var/canMouseDown = FALSE
+	
+	//item hover FX
```
Comment (Mothblocks) [inline code/game/objects/items.dm:195]:
```suggestion
	/// item hover FX
```

Doc comment.

File: code/game/objects/items.dm [hunk 3]
```diff
@@ -829,11 +832,54 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
 		var/timedelay = usr.client.prefs.tip_delay/100
 		var/user = usr
 		tip_timer = addtimer(CALLBACK(src, .proc/openTip, location, control, params, user), timedelay, TIMER_STOPPABLE)//timer takes delay in deciseconds, but the pref is in milliseconds. dividing by 100 converts it.
+		var/mob/living/L = usr
```
Comment (Mothblocks) [inline code/game/objects/items.dm:835]: Better name.

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/objects/items.dm [hunk 4]
```diff
@@ -829,11 +832,54 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
...
+/obj/item/MouseDrop(atom/over, src_location, over_location, src_control, over_control, params)
+	. = ..()
+	remove_outline()
 
 /obj/item/MouseExited()
 	deltimer(tip_timer)//delete any in-progress timer if the mouse is moved off the item before it finishes
 	closeToolTip(usr)
+	remove_outline()
 
+/obj/item/proc/apply_outline(colour = null)
```
Comment (Mothblocks) [inline code/game/objects/items.dm:850]: `colour` is a confusing name, as it constantly conflicts with `color`.

Make this `outline_color`, or whatever it is.

File: code/game/objects/items.dm [hunk 5]
```diff
@@ -829,11 +832,54 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
...
+	if(!(item_flags & IN_INVENTORY || item_flags & IN_STORAGE) || QDELETED(src) || isobserver(usr))
+		return
+	var/theme = lowertext(usr.client.prefs.UI_style)
+	if(!colour)
+		switch(theme) //AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
```
Comment (Mothblocks) [inline code/game/objects/items.dm:855]: It's fine, I think we need to keep our UI as just the basic and glass ones anyway, and change their colors. That's for the future.

File: code/game/objects/items.dm [hunk 6]
```diff
@@ -829,11 +832,54 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
...
+			if("retro")
+				colour = COLOR_THEME_RETRO
+			if("slimecore")
+				colour = COLOR_THEME_SLIMECORE
+			if("operative")
+				colour = COLOR_THEME_OPERATIVE
+			if("clockwork")
+				colour = COLOR_THEME_CLOCKWORK
+			if("glass")
+				colour = COLOR_THEME_GLASS
```
Comment (Mothblocks) [inline code/game/objects/items.dm:869]: Glass theme has customizable color. Use something that uses that instead.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/game/objects/items.dm (hunk 6 above)
Comment (maxymax13) [reply code/game/objects/items.dm:869]: where is that pref for that set ingame, and where is it stored

File: code/game/objects/items.dm (hunk 3 above)
Comment (maxymax13) [reply code/game/objects/items.dm]: L retracted

File: code/game/objects/items.dm (hunk 4 above)
Comment (maxymax13) [reply code/game/objects/items.dm]: british exterminated

File: code/game/objects/items.dm (hunk 6 above)
Comment (Mothblocks) [reply code/game/objects/items.dm]: I don't know off hand, shouldn't be hard to find yourself though.

File: code/game/objects/items.dm [hunk 7]
```diff
@@ -832,48 +832,48 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
 		var/timedelay = usr.client.prefs.tip_delay/100
 		var/user = usr
 		tip_timer = addtimer(CALLBACK(src, .proc/openTip, location, control, params, user), timedelay, TIMER_STOPPABLE)//timer takes delay in deciseconds, but the pref is in milliseconds. dividing by 100 converts it.
-		var/mob/living/L = usr
-		if(istype(L) && L.incapacitated())
-			apply_outline(COLOR_RED_GRAY)
+		var/mob/living/alive = usr
```
Comment (Mothblocks) [inline code/game/objects/items.dm:835]: This is still a bad variable name, it doesn't describe the context.

Just name it `user` or `living_user`.

File: code/game/objects/items.dm [hunk 8]
```diff
@@ -832,48 +832,48 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
...
+		if(istype(alive) && alive.incapacitated()) // are we alive? are we handcuffed?
```
Comment (Mothblocks) [inline code/game/objects/items.dm:836]: This is a bad comment. Comments should say why, not what.

File: code/game/objects/items.dm [hunk 9]
```diff
@@ -832,48 +832,48 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
...
+			apply_outline(COLOR_RED_GRAY) //red meeans stop trying to click on it
 		else
 			apply_outline()
 
 /obj/item/MouseDrop(atom/over, src_location, over_location, src_control, over_control, params)
 	. = ..()
-	remove_outline()
+	remove_outline() //remove the outline when we drag and drop this
```
Comment (Mothblocks) [inline code/game/objects/items.dm:843]: Same with this, this is a bad comment.

**Comment by maxymax13** [comment]
just got trolled by a workflow fail email notification in a line that i already fixed 😢

**Review by Mothblocks — APPROVED** [review]
LGTM, undraft if ready.

**Comment by maxymax13** [comment]
i'll add the coloring and the disable pref and then i think that'll be everything

File: code/game/objects/items.dm (hunk 6 above)
Comment (maxymax13) [reply code/game/objects/items.dm]: i've looked everywhere in all 500 separate ingame menus where preferences are hidden, through all the user preferences and hud files and found absolutely nothing about any hud recoloring feature

Comment (Mothblocks) [reply code/game/objects/items.dm]: I can't seem to find it either, so I guess it wasn't merged. 🤷

**Comment by Mothblocks** [comment]
don't web edit behavior like that nooo

just test in game

**Comment by maxymax13** [comment]
it works 🤷 

https://user-images.githubusercontent.com/38844529/110873216-02565780-82b0-11eb-9467-58fe5cc841a5.mp4

File: code/modules/client/preferences_savefile.dm [hunk 10]
```diff
@@ -246,6 +247,7 @@ SAVEFILE UPDATING/VERSIONING - 'Simplified', or rather, more coder-friendly ~Car
 	parallax = sanitize_integer(parallax, PARALLAX_INSANE, PARALLAX_DISABLE, null)
 	ambientocclusion	= sanitize_integer(ambientocclusion, FALSE, TRUE, initial(ambientocclusion))
 	screentip_pref	= sanitize_integer(screentip_pref, FALSE, TRUE, initial(screentip_pref))
+	itemoutline_pref	= sanitize_integer(itemoutline_pref, FALSE, TRUE, initial(itemoutline_pref))
```
Comment (Mothblocks) [inline code/modules/client/preferences_savefile.dm:250]:
```suggestion
	itemoutline_pref = sanitize_integer(itemoutline_pref, FALSE, TRUE, initial(itemoutline_pref))
```

I know the others are like this. They shouldn't be.

**Review by Mothblocks — APPROVED** [review]

**Pull request merged**